### PR TITLE
refactor: _lfortran_string_read_i32

### PR DESCRIPTION
**Pull request by sriganeshres**
towards: #8241

### CODE REVIEW

File: src/libasr/runtime/lfortran_intrinsics.c [hunk 1]
```diff
@@ -5082,19 +5082,35 @@ LFORTRAN_API void _lfortran_string_write(char **str_holder, bool is_allocatable,
...
+    strncpy(tmp, str, len);
+    tmp[len] = '\0';
+    sscanf(tmp, format, i);
+    free(tmp);
 }
 
 LFORTRAN_API void _lfortran_string_read_f32(char *str, int64_t len, char *format, float *f) {
-    sscanf(str, format, f);
+    char* tmp = (char*)malloc(len + 1);
+    strncpy(tmp, str, len);
```
Comment (assem2002) [inline src/libasr/runtime/lfortran_intrinsics.c]: Don't we have any other way to handle this function without using C-style strings?

Comment (sriganeshres) [reply src/libasr/runtime/lfortran_intrinsics.c]: We can but we need to use a bounded writer. Do you want me to do that.

Comment (sriganeshres) [reply src/libasr/runtime/lfortran_intrinsics.c]: I am doing that.

Comment (assem2002) [reply src/libasr/runtime/lfortran_intrinsics.c]: Can you explain further? I tried to pass the problem to some GPTs and obviously the options out there aren't different (performance wise) from what we have here.

File: src/libasr/runtime/lfortran_intrinsics.c [hunk 2]
```diff
@@ -5082,19 +5082,35 @@ LFORTRAN_API void _lfortran_string_write(char **str_holder, bool is_allocatable,
 }
 
 LFORTRAN_API void _lfortran_string_read_i32(char *str, int64_t len, char *format, int32_t *i) {
-    sscanf(str, format, i);
+    char* tmp = (char*)malloc(len + 1);
+    strncpy(tmp, str, len);
```
Comment (assem2002) [inline src/libasr/runtime/lfortran_intrinsics.c]: Can you please use `memcpy`?, as `strncpy` has implication that it stops whenever encounters a `\0` and Fortran strings can have such characters as part of the string.

File: src/libasr/runtime/lfortran_intrinsics.c [hunk 3]
```diff
@@ -5082,19 +5106,165 @@ LFORTRAN_API void _lfortran_string_write(char **str_holder, bool is_allocatable,
...
+    int sign = 1;
+    if (p < end && *p == '-') { sign = -1; p++; }
+    else if (p < end && *p == '+') { p++; }
+
+    int32_t value = 0;
+    while (p < end && isdigit((unsigned char)*p)) {
+        value = value * 10 + (*p - '0');
+        p++;
+    }
+    *i = value * sign;
```
Comment (certik) [inline src/libasr/runtime/lfortran_intrinsics.c]: I don't like this. From here and below we are expanding a single line `sscanf` into 20+ lines, copy & pasting the same code (possibly with some tiny changes --- hard to tell) all over.

My first question is: what is the motivation, is sscanf not enough for what we need? Is there some Fortran test case that didn't work in main, but now works? If so, the test should be added in this PR.

If the answer is that it is not good enough, then we need to refactor all this code into functions to parse integers, don't repeat it. And then we need to add tests for these parsing routines, as it is easy to make mistakes in them, and forget about some corner case, like an empty string, invalid string, etc. The best way to test it is to call the runtime from our `ctest` infrastructure and just add tests specifically for the parsing. Then use those well-tested functions here.

Comment (sriganeshres) [reply src/libasr/runtime/lfortran_intrinsics.c]: No I wanted to use non c style strings that why i mdid not make any tmp C style strings and made it to mimic sscanf as asked here https://github.com/lfortran/lfortran/pull/8289#discussion_r2262818697. If you dont want this and just change `strncpy` to `memcpy` I can do that.

Comment (sriganeshres) [reply src/libasr/runtime/lfortran_intrinsics.c]: No problem i have changed `strncpy` to `memcpy`

File: src/libasr/runtime/lfortran_intrinsics.c [hunk 4]
```diff
@@ -5082,19 +5106,40 @@ LFORTRAN_API void _lfortran_string_write(char **str_holder, bool is_allocatable,
 }
 
 LFORTRAN_API void _lfortran_string_read_i32(char *str, int64_t len, char *format, int32_t *i) {
-    sscanf(str, format, i);
+    char *buf = (char*)malloc(len + 1);
+    if (!buf) return; // handle allocation failure if needed
```
Comment (certik) [inline src/libasr/runtime/lfortran_intrinsics.c]: What is the purpose of this comment --- either handle it, or explain why it doesn't need to be handled.

Comment (sriganeshres) [reply src/libasr/runtime/lfortran_intrinsics.c]: it is just a malloc error. Comment is nothing we can ignore that.

File: src/libasr/runtime/lfortran_intrinsics.c [hunk 5]
```diff
@@ -5082,19 +5106,40 @@ LFORTRAN_API void _lfortran_string_write(char **str_holder, bool is_allocatable,
...
 }
 
 LFORTRAN_API void _lfortran_string_read_f32(char *str, int64_t len, char *format, float *f) {
-    sscanf(str, format, f);
+    char *buf = (char*)malloc(len + 1);
+    if (!buf) return;
+    memcpy(buf, str, len);
+    buf[len] = '\0';
+    sscanf(buf, format, f);
+    free(buf);
```
Comment (certik) [inline src/libasr/runtime/lfortran_intrinsics.c:5133]: This should be completely redesigned: https://github.com/lfortran/lfortran/issues/8310, but let's do it in a subsequent PR.

File: src/libasr/runtime/lfortran_intrinsics.c (hunk 4 above)
Comment (certik) [reply src/libasr/runtime/lfortran_intrinsics.c]: Then remove it. Don't add comments that are "nothing".

Comment (certik) [inline src/libasr/runtime/lfortran_intrinsics.c]:
```suggestion
    if (!buf) return;
```
